WT-14393 Implement new crashpoints controlled by debug.checkpoint_crash_point config

raviprakashgiri29 opened the pull request:
This pull request implements two new crash points (after the log flush and after metadata sync) controlled by the debug.checkpoint_crash_point configuration to simulate and investigate crashes happening after the checkpoint log flush.

pmacko86 commented on src/checkpoint/checkpoint_txn.c:
```diff
@@ -1535,6 +1537,10 @@ __checkpoint_db_internal(WT_SESSION_IMPL *session, const char *cfg[])
 
+    /* Crash before checkpoint stop if checkpoint crash point is configured. */
+    if (ckpt_crash_point_after_ckpt_files == CKPT_CRASH_BEFORE_CKPT_STOP)
+        __wt_debug_crash(session);
```
Does this happen after the turtle file is updated?

We should not have any crash points after the turtle file update, because as far as the on-disk state is concerned, the checkpoint is complete when the turtle file is updated. We should keep the semantics of these crash points so that the checkpoint is incomplete and should be ignored if we do trigger the crash.

etienneptl replied:
I think the reason here is because we haven't logged the end of the checkpoint yet in the log which recovery depends on.

raviprakashgiri29 replied:
The last crash point was located after the turtle file update, which was the cause of the failure. I've refactored the code and removed this final crash point.

pmacko86 commented on src/checkpoint/checkpoint_txn.c:
```diff
@@ -1251,18 +1252,18 @@ __checkpoint_db_internal(WT_SESSION_IMPL *session, const char *cfg[])
-        ckpt_relative_crash_step = ((u_int)ckpt_crash_point / (WT_THOUSAND / ckpt_total_steps));
-
-        if (ckpt_relative_crash_step < session->ckpt.handle_next)
+        ckpt_total_crash_points = session->ckpt.handle_next + 2;
```
Is `+ 2` correct? I thought this number should match the number of the `enum` values, which is 3?

Please also add a comment about where this number is coming from to ensure that anyone working on this in the future knows how to update it.

etienneptl replied:
We have can a new entry "enum_end" to make this more robust.

pmacko86 replied:
Maybe you could make this clearer by changing the comment to something like this?
```
        /*
         * Calculate total checkpoint crash points. The total checkpoint points required are the
         * number of data handles that need to be checkpointed plus the additional crash points,
         * which is the number of values in the ckpt_crash_state enum. 
         */
```

But I do also like Etienne's idea—I'm assuming he meant making the enum to be something like this?
```
    enum {
        CKPT_CRASH_BEFORE_METADATA_SYNC = 0,
        CKPT_CRASH_BEFORE_METADATA_UPDATE,
        CKPT_CRASH_ENUM_END
    } ckpt_crash_state;
```
and then doing this here?
```
ckpt_total_crash_points = session->ckpt.handle_next + CKPT_CRASH_ENUM_END;
```

@etienneptl, is this what you had in mind? If so, do you know if it works with the C standard, or if there is a danger that this is compiler-specific?

etienneptl replied:
Yes that's what I had in mind, I just thought of it because I have seen this in the code before so I thought about it "from experience". I do not know if it is usually the recommended way or comply with the C standard. Worth double checking @raviprakashgiri29 !

pmacko86 commented on src/checkpoint/checkpoint.h:
```diff
@@ -23,6 +23,11 @@ struct __wt_ckpt_session {
 
     /* Checkpoint crash. */
     u_int crash_point; /* Crash point in the middle of checkpoint process */
+    enum {
```
I like that you made it an `enum`!

pmacko86 reviewed (commented):
Almost there! I like the overall shape of the solution; just please verify that the last crash doesn't happen after the turtle file is updated. As explained in the comment, we shouldn't be crashing that late.

raviprakashgiri29 commented:
> Almost there! I like the overall shape of the solution; just please verify that the last crash doesn't happen after the turtle file is updated. As explained in the comment, we shouldn't be crashing that late.

Thanks @pmacko86. The last crash point was located after the turtle file update, which was the cause of the failure. I've refactored the code and removed this final crash point.

pmacko86 reviewed (approved):
LGTM. Thank you!
